File: recursion/fib.py
# 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ... 

# recurssive approach 
# O(2^n) time | O(n) space
def fib(n):
    if n == 1:
        return 0
    elif n == 2:
        return 1
    else:
        return fib(n - 1) + fib(n - 2)

# dynamic programming approach (bottom up)
# O(n) time | O(n) space
def fib_dp(n, cache = {1: 0, 2: 1}):
    if n in cache:
        return cache[n]
    else:
        cache[n] = fib_dp(n - 1, cache) + fib_dp(n - 2, cache)
        return cache[n]

# iterative solution
# O(n) time | O(1) space
def fib_iterative(n):
    lastTwo = [0, 1]
    counter = 3
    while counter <= n:
        nextFib = lastTwo[0] + lastTwo[1]
        lastTwo[0] = lastTwo[1]
        lastTwo[1] = nextFib
        counter += 1
    if n > 1:
        return lastTwo[1]
    else:
        return lastTwo[0]


def test():
    n = 5
    print('the', n, 'th fib number is: ', fib_iterative(n))

test()

File: js/recursion/fib.js
/**
 * time-complexity:
 *  -O(2^n) - height of recursive stack
 * 
 * space-complexity:
 *  -O(n)
 * 
 * @param {int} n - function returns the n-th fibinacci number 
 */
function fib(n) {
    if (n === 1) {
        return 0;
    } else if (n === 2) {
        return 1
    } else {
        return fib(n - 1) + fib(n - 2);
    }
}

/**
 * time-complexity:
 *  -O(n)
 * 
 * space-complexity:
 *  -O(n)
 * 
 * @param {int} n - function returns the n-th fibinacci number 
 */
function fib_dp(n, cache = { 1: 0, 2: 1 }) {
    if (n in cache) {
        return cache[n];
    } else {
        cache[n] = fib_dp(n - 1, cache) + fib_dp(n - 2, cache);
        return cache[n]
    }
}


/**
 * time-complexity:
 *  -O(n)
 * 
 * space-complexity:
 *  -O(1)
 * 
 * @param {int} n - function returns the n-th fibinacci number 
 */
function fib_iterative(n) {
    let last_two = [0, 1];
    let counter = 3;

    while (counter <= n) {
        let nextFib = last_two[0] + last_two[1];
        last_two[0] = last_two[1];
        last_two[1] = nextFib;
        counter = counter + 1;
    }
    if (n > 1) {
        return last_two[1]
    } else {
        return last_two[0]
    }
}



function test() {
    console.log('8th fib recursively: ', fib(8))
    console.log('8th fib dp: ', fib_dp(8))
    console.log('8th fib dp: ', fib_iterative(8))
}

test()